### ISSUE TITLE
Are partial firmware updates supported?

### ISSUE THREAD

**Issue by crtag** [Lora-net/lr1110_updater_tool#1]
I'm having a problem updating the firmware as we did not create any hardware options to connect to LR1110 directly, nor our host MCU has enough memory to hold the entire LR1110 firmware. All I have is SWD interface with MCU and I was hoping that I can update firmware by running two executables on host MCU holding half of the LR1110 firmware. The offset for the second half is respected, and 256 byte long chunks too in the first half. After running through that process LR1110 seems to be stuck in the bootloader mode. Is this approach valid?

**Comment by matthieuantoine** [comment]
Partial firmware update as described in your message actually work.

I did a simple test in the current version of this tool (line 131 in lr1110_firmware_update.c):

```c
        lr1110_bootloader_write_flash_encrypted_full( radio, 0, buffer, length );
```

replaced by:

```c
        lr1110_bootloader_write_flash_encrypted_full( radio, 0, buffer, 64 );
        system_gpio_wait_for_state( lr1110_busy, 0 );
        lr1110_system_reset( radio );
        lr1110_bootloader_write_flash_encrypted_full( radio, 64 * 4, buffer + 64, length - 64 );
```

Several points to pay attention to:
- Make sure you wait for the BUSY line to go low before you load the next executable - writing data in flash takes time and interrupting the process will corrupt the memory
- Note that the parameter `offset` is byte-wise while the parameter `length` is word-wise.

**Comment by crtag** [comment]
@matthieuantoine thanks for the note, I will try to see if the issue is in HAL layer then, I'm using STM32Cube generated code so I had to define SPI interaction following the example in this repo.
Regarding BUSY state - in both this repo it's being waited on at the lower level in lr1110_hal.c line 75
`system_gpio_wait_for_state( radio_local->busy, SYSTEM_GPIO_PIN_STATE_LOW );`
so my impression it's not necessary to do this on a higher level, or am I missing something essential?

I did notice for the testing you issue hardware reset via `lr1110_system_reset( radio );` - does it actually mean that as long as LR1110 is in bootloader mode you can flash chunks of code and then, every time it's getting reset the bootloader will verify the signature and will take it out of the bootloader only if the firmware checks out properly? My situation is that I literally have to reboot the whole system to upload the second chunk, and I thought that could confuse the bootloader.

**Update:**
I've carefully repeated the experiment, and there's some unexpected behaviour:

- I've changed the way LR1110 is getting into bootstrap mode - now it's done the hardware way via RESET+BUSY as documented;
- This effectively brings LR1110 into bootstrap, I can see the version object as 
```
hw=0x21
type=0xdf
fw=0x6500
```
and it seems to be valid

- Erasing (takes around 2 secs unit BUSY gets cleared)
- Writing the first chunk of the firmware w/o any SPI errors
- Uploading the second app with the rest of the firmware, effectively rebooting the system;
- Checking the version - it's as follows:
```
hw=0x14
type=0x00
fw=0x0000
```
This is not quite expected!
- I made sure to take LR1110 into bootloader mode via `lr1110_bootloader_reboot( &radio, true );`
- Version checks out fine now
- Second chunk of firmware is uploaded w/o any SPI errors
- Reset (via hardware) LR1110 
- Request version via bootstrap command. 
At this point in time, I find BUSY is high, and time doesn't cure it.

I can take it from BUSY state two ways:
- Hardware bootstrap procedure, works fine and bootstrap commands are working so I'm able to re-try anything;
- Take NSS low on SPI, but it lasts only until I send another command via SPI.

I've noticed this `hw=0x14` was happening earlier in my other experiments, so I'm not sure what it indicates. 
However, BUSY state seems to tell something is corrupted, but why it keeps this state, and nothing, even hardware RESET, doesn't help.

**Update #2:**
Previously, I've been trying to upgrade to the modem, using the provided header file.
Just to give it another go, I tried it with the transceiver 0x0304.
After going through the above steps it returned the following:

running `lr1110_bootloader_get_version()` returned:
```
hw=0x21
type=0x01
fw=0x0304
```
running 
`lr1110_system_get_version()` returned:

```
hw=0x13
type=0x00
fw=0xC000
```

Which seems to be a successful upgrade. Not so sure regarding the `lr1110_system_get_version()` result, though.

However, I can confirm that hardware reset after the first chunk upload brought it into this state as well:
```
hw=0x14
type=0x00
fw=0x0000
```

**Could it be the issue with the GitHub modem firmware file?**


**Update #3:**
I've downloaded binaries from here https://github.com/Lora-net/lr1110_updater_tool/wiki
and used them on eval kit for LR1110 I have. 
Using modem update rendered the kit useless with the same symptoms as above - the upgrade never finishes and seems to stuck on line where it attempts to get the modem version.
Flashing it with the transceiver version works fine.

I suspect there's some undocumented LR1110 state from looking at this:
https://github.com/Lora-net/lr1110_evk_demo_app/blob/25d6a647013488363b872ec451171831031e4df3/embedded/application/src/main.cpp#L244

That is BUSY high is expected state of modem after reset, but I'm wondering what does it mean and how to deal with that.

@matthieuantoine that would be great if any docs related to this would be provided. Thanks in advance!

**Comment by crtag** [comment]
It appears that most of the things I reported are due to tiny details hidden across separate pieces of documentation that are marked as "notes" where those should be in huge red text "Important" boxes.

To summarise:

- Partial updates are supported as described above, however, it's **essential to ensure the LR1110 is in bootloader mode** before writing another chunk of firmware! I.e. the system does not automatically revert into bootloader mode if the firmware is corrupted - I did not observe this behaviour with the version of chips I have.
- Functional polarity of the **BUSY** signal is opposite in TX and Modem modes of the system. One of the things that should be in the red "Important" box in docs.

yeah, RTFM. Just a bit disappointing some key things not outlined in a single doc.

One issue still stands - using the binary for the current version did not go through modem update as it's expected with the eval kit.

**Comment by matthieuantoine** [comment]
Sorry for the lack of clarity in the documentation - we'll make changes to improve it.

Concerning the points you raised:
- The LR1110 does not jump into the firmware if this one is "corrupted" (i.e. the signature check does not return OK) - it stays in bootloader mode after POR or reset. It is the case if you are in the middle of an update
- What do you mean by "Functional polarity of the BUSY signal is opposite in TX and Modem modes of the system"?

Regarding the issue that still stands, can you please share more details?
- What is the binary file you are using?
- Is there something that went wrong with the drag-and-drop process?

**Comment by crtag** [comment]
@matthieuantoine Just to clarify my points and respond to your notes:

- LR1110 does not jump into the firmware (it's corrupted, of course), neither it automatically jumps into bootloader mode if I'm in the middle of an update, ie the firmware is "corrupted". If LR1110 version is requested via bootloader command at this stage the following is responded:

```
hw=0x14
type=0x00
fw=0x0000
```
I have observed this a few times and, apparently, this was the blocker to the process as I was assuming resetting LR1110 the hardware way would take it into the bootloader mode automatically as specified. To overcome this I had to make an extra `lr1110_bootloader_reboot( &radio, true );` call. Works like a charm after that.

- Functional polarity of the *BUSY* signal gets mentioned only in section 2.2 of the modem manual as follows:
> Note: The BUSY signal has inverted polarity in the Transceiver and Modem implementations.

which also becomes obvious when you start looking at SPI diagrams. 

Since chip default mode is TX and one starts working with that mode, it's easy to trip over on that BUSY polarity which needs to be taken into account when providing SPI HAL layer specific to host MCU.

- I've taken provided images from here: https://github.com/Lora-net/radio_firmware_images/tree/master/lr1110, also have built a custom binary with updated logic, but to no avail. I'd suggest the problem on my end, I had to focus on our device rather than trying hard enough with the eval kit.
- Used STM32 programmer, verified upload etc, so it wasn't a drag-n-drop. I may give it a few more careful tries later if this is important.

There're other questions rather than reports, is there a way to ask those via some direct channels? Mostly about recommended app architecture, as I'm looking at eval kit and demo repos, and they have a bit different approach to handling the interrupts (no software handling pretty much).
